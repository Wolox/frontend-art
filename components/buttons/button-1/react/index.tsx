import React from 'react';

import '../styles.scss';

function Button1() {
  return (
    <div className="column center btn-group-1">
      <button className="btn btn-1 m-bottom-2" disabled={true}>
        <div className="btn__content">cancel</div>
      </button>
      <button type="button" className="btn btn-1">
        <div className="btn__content">submit</div>
      </button>
    </div>
  );
}

export default Button1;
