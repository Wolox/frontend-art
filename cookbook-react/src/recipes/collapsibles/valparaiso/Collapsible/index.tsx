import React, { ReactNode, useState, useRef, useEffect } from 'react';
import cn from 'classnames';

import styles from './styles.module.scss';

interface CollapsibleProps {
  children: ReactNode;
  title: string;
  className?: string;
  onChange?: (isOpen: boolean) => void;
  defaultIsClosed?: boolean;
  collapsibleId: string;
}

const useCollapse = (defaultIsOpen: boolean, onCollapse?: (isOpen: boolean) => void) => {
  const [collapsed, setCollapsed] = useState(!defaultIsOpen);
  const collapsibleRef = useRef<HTMLDivElement>(null);

  const setCollapsedValue = (height: number, isCollapsed: boolean) => {
    const heightString = isCollapsed ? '0px' : `${height}px`;
    collapsibleRef.current!.style.height = heightString;
  };

  useEffect(() => {
    const calculateHeight = () => {
      const maxHeight = collapsibleRef.current?.scrollHeight ?? 0;
      setCollapsedValue(maxHeight, collapsed);
    };

    calculateHeight();
    window.addEventListener('resize', calculateHeight);
    return () => {
      window.removeEventListener('resize', calculateHeight);
    };
  }, [collapsed]);

  const handleCollapse = () => {
    setCollapsed(!collapsed);
    setCollapsedValue(collapsibleRef.current?.scrollHeight!, !collapsed);
    onCollapse?.(collapsed);
  };

  return { handleCollapse, collapsibleRef, collapsed };
};

function Collapsible({
  collapsibleId,
  title,
  className,
  defaultIsClosed = false,
  children,
  onChange
}: CollapsibleProps) {
  const { handleCollapse, collapsibleRef, collapsed } = useCollapse(!defaultIsClosed, onChange);

  return (
    <div className={cn(className, styles.collapsibleContainer, 'column')}>
      <div className="row space-between m-bottom-2">
        {title && <h3 className={styles.title}>{title}</h3>}
        <button
          type="button"
          className={cn(styles.icon, { [styles.iconClicked]: collapsed })}
          aria-expanded={!collapsed}
          aria-controls={collapsibleId}
          onClick={handleCollapse}
        />
      </div>
      <div id={collapsibleId} ref={collapsibleRef} className={cn(styles.content)} role="region">
        {children}
      </div>
    </div>
  );
}

export default Collapsible;
