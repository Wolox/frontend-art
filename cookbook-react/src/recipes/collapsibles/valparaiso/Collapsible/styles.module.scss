$background-color: #FFF;
$border-color: #848486;
$title-color: #08080D;
$button-color: #00B876;
$button-color-hover: #04A36A;

$icon-height: 12px;
$icon-width: 18px;

.collapsible-container {
  background-color: $background-color;
  border: 1px solid $border-color;
  border-radius: 10px;
  padding: 20px;
}

.content {
  overflow: hidden;
  transition: height 0.3s ease-out;
  transform-origin: top;
}

.title {
  color: $title-color;
  font-size: 32px;
  font-weight: 700;
  line-height: 37px;
}

.icon {
  align-self: center;
  height: $icon-height;
  position: relative;
  transition: transform 0.3s linear;
  width: $icon-width;

  &::before,
  &::after {
    border: calc(#{$icon-width} / 2) solid transparent;
    content: '';
    display: block;
    position: absolute;
  }

  &::before {
    border-top-color: $border-color;
  }

  &::after {
    border-top-color: $background-color;
    top: 4px;
  }
}

.icon-clicked {
  transform: rotate(180deg);
}
