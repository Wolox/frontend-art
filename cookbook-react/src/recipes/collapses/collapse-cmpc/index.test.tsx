import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';

import Collapsible from './index';

describe('Collapsible component', () => {
  const title = 'Collapsible Title';
  const className = 'collapsible-class';
  let fragment: DocumentFragment | null = null;
  let collapsibleContainer: Element | null = null;

  beforeEach(() => {
    const { asFragment, container } = render(
      <Collapsible title={title} isCollapsed className={className}>
        <div> some content </div>
      </Collapsible>
    );

    fragment = asFragment();
    collapsibleContainer = container;
  });

  it('renders without crashing', () => {
    expect(fragment).toMatchSnapshot();
    expect(screen.getByText(title)).toBeInTheDocument();
    expect(collapsibleContainer.firstChild).toHaveClass(className);
  });

  it('toggle between show and hide the children content', async () => {
    expect(screen.queryByText('some content')).not.toBeInTheDocument();

    fireEvent.click(screen.getByText(title));

    expect(await screen.findByText('some content')).toBeInTheDocument();
  });
});
