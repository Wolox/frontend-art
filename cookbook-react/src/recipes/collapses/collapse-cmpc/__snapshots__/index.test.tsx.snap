// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapsible component renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="collapsible-class collapsibleContainer column"
  >
    <div
      class="row space-between"
    >
      <span
        class="title"
      >
        Collapsible Title
      </span>
      <div
        class="row middle"
      >
        <button
          class="m-right-10 button"
          type="button"
        >
          Action
        </button>
        <button
          class="icon iconClicked"
          type="button"
        />
      </div>
    </div>
    <div
      class="content collapsedContent"
    />
  </div>
</DocumentFragment>
`;
